src/datarobotx/idp/common/path_dataset.py

```diff
@@ -0,0 +1,154 @@
+#
+# Copyright 2024 DataRobot, Inc. and its affiliates.
+#
+# All rights reserved.
+#
+# DataRobot, Inc.
+#
+# This is proprietary source code of DataRobot, Inc. and its
+# affiliates.
+#
+# Released under the terms of DataRobot Tool and Utility Agreement.
+# https://www.datarobot.com/wp-content/uploads/2021/07/DataRobot-Tool-and-Utility-Agreement.pdf
+
+from copy import deepcopy
+import pathlib
+from pathlib import Path, PurePosixPath
+import shutil
+import sys
+import tempfile
+from typing import Any, Dict, Optional, Union
+import weakref
+
+try:
+    import fsspec
+    from kedro.io.core import (
+        AbstractVersionedDataset,
+        DatasetError,
+        Version,
+        get_filepath_str,
+        get_protocol_and_path,
+    )
+except ImportError as e:
+    raise ImportError("Consider including kedro in your project requirements`") from e
+
+
+# https://discuss.python.org/t/fixing-subclassing-in-pathlib/8983/16
+if sys.platform == "win32":
+    _PathBase = pathlib.WindowsPath
+else:
+    _PathBase = pathlib.PosixPath
+
+
+class WeakReferencablePath(_PathBase):
+    """weakref-friendly path.
+
+    pathlib.Path, like many built-ins, can't be used by default with weakref.
+    """
+
+    pass
+
+
+class PathDataset(AbstractVersionedDataset):  # type: ignore
+    """Kedro connector for interacting with arbitrary files and directories.
+
+    Accommodates full directories as well as  data that is not ideal to
+    serialize via pickle, json or other tabular formats (e.g. huggingface
+    cache dir, directory of PDFs, audio files or video files)
+
+    Transfers the provided pathlib.Path or tempfile.TemporaryDirectory
+    to `filepath` using the normal fsspec mechanisms.
+
+    When loading the path dataset as a single file, a pathlib.Path is
+    returned for that file.  When loading the path dataset as a
+    directory of files, a temporary directory containing the dataset is passed
```

Comment:
this paragraph is a bit unclear, maybe something like "When loading, a temporary directory is created locally containing the requested directory or file and pathlib.Path to this directory or file is returned. The temporary directory is cleaned up when the Path is garbage collected."